languagetool-language-modules/ar/src/main/java/org/languagetool/rules/ar/ArabicPunctuationWhitespaceRule.java

```diff
@@ -0,0 +1,228 @@
+        && !containsDigit(prevPrevToken)
+        && !containsDigit(token)
+        && !getCommaCharacter().equals(prevPrevToken)
+        && !getCommaCharacterArabic().equals(prevPrevToken)) {
+        msg = messages.getString("missing_space_after_comma");
+        suggestionText = getCommaCharacterArabic() + " " + tokens[i].getToken();
+      } else if (prevWhite) {
+        if (isRightBracket(token)) {
+          boolean isException = token.equals("]") && prevToken.equals(" ") && prevPrevToken.equals("["); // "- [ ]" syntax e.g. on GitHub
+          if (!isException) {
+            msg = messages.getString("no_space_before");
+            suggestionText = token;
+          }
+        } else if (token.equals(getCommaCharacter()) || token.equals(getCommaCharacterArabic())) {
+          msg = messages.getString("space_after_comma");
+          suggestionText = getCommaCharacterArabic();
+          // exception for duplicated comma (we already have another rule for that)
+          if (i + 1 < tokens.length && (getCommaCharacter().equals(tokens[i + 1].getToken()) || getCommaCharacterArabic().equals(tokens[i + 1].getToken()))) {
+            msg = null;
+          }
+          if (i + 1 < tokens.length && !tokens[i + 1].isWhitespace()) {
+            suggestionText = getCommaCharacterArabic() + " ";
+          }
+        } else if (token.equals(".") && !isDomain(tokens, i + 1) && !isFileExtension(tokens, i + 1)) {
+          msg = messages.getString("no_space_before_dot");
+          suggestionText = ".";
+          // exception case for figures such as ".5" and ellipsis
+          if (i + 1 < tokens.length && isDigitOrDot(tokens[i + 1].getToken())) {
+            msg = null;
+          } else if (i + 2 < tokens.length && tokens[i + 1].getToken().equals("/") && tokens[i + 2].getToken().matches("[a-zA-Z]+")) {
+            // commands like "./validate.sh"
+            msg = null;
+          }
+        } // arabic question mark
+        else if (token.equals("؟") || token.equals("?")) {
+          msg = messages.getString("no_space_before_dot");
+          suggestionText = "؟";
+          // exception case
+        } // arabic semi colon
+        else if (token.equals("؛") || token.equals(";")) {
+          msg = messages.getString("no_space_before_dot");
+          suggestionText = ";";
+          // exception case
+        }
+      }
+      if (msg != null && !isException(tokens, i)) {
+        int fromPos = tokens[i - 1].getStartPos();
+        int toPos = tokens[i].getEndPos();
+        RuleMatch ruleMatch = new RuleMatch(this, sentence, fromPos, toPos, msg);
+        ruleMatch.setSuggestedReplacement(suggestionText);
+        ruleMatches.add(ruleMatch);
+      }
+      prevPrevToken = prevToken;
+      prevToken = token;
+      prevWhite = isWhitespace && !tokens[i].isFieldCode(); // LO/OO code before comma/dot
+    }
+
+    return toRuleMatchArray(ruleMatches);
+  }
+
+  private boolean isDomain(AnalyzedTokenReadings[] tokens, int i) {
```

Comment:
There seems to be a lot of code duplication with `CommaWhitespaceRule`, could this be avoided?

Reply:
This is because the `match` function of `CommaWhitespaceRule` is **final**.

Reply:
May be if we duplicate rules with different type of commas.

Reply:
Feel free to remove the `final` in `CommaWhitespaceRule.match()` if it helps to avoid duplication.

Reply:
Actually I prefer the separate rules solution. It seems to be cleaner.